src/modules/ctc/src/ctc.cc

```diff
@@ -184,6 +231,36 @@ types::Error Ctc::GetTrainPointerById(const types::TrainId train_id, std::shared
     return error;
 }
 
+types::Error Ctc::GetTrainPointerFromBlockOccupancy(const types::BlockId block_id, std::shared_ptr<ctc::Train> &train_pointer)
+{
+    types::Error error = types::Error::ERROR_INVALID_TRAIN;
+    for (ctc::Train &train : train_schedules_)
+    {
+        if (train.authority.front() == block_id)
+        {
+            train_pointer = std::shared_ptr<ctc::Train>(&train, [](ctc::Train *) {
+                });
+            error = types::Error::ERROR_NONE;
```

Comment:
You can add a `break` statement to terminate the for loop early after a train is found.  This applies to the `GetBlockPoitnerById` method as well.

Reply:
removed and replaced with find_if